src/Einsum.jl

```diff
@@ -27,336 +28,404 @@ macro einsum_checkinbounds(ex)
-        $ex_create_arrays
+        $type_definition
+        $output_definition
+        $(dimension_checks...)
         let
-            $(ex_check_dims...)
-            $ex_get_type
-            $ex
+            $expr
         end
-        $(lhs_arr[1])
+        $(lhs_arrays[1])
     end
 
     return esc(full_expression)
 end
 
+function check_index_occurrence(lhs_indices, rhs_indices)
+    if !issubset(lhs_indices, rhs_indices)
+        missing_indices = setdiff(lhs_indices, rhs_indices)
+
+        if length(missing_indices) == 1
+            missing_string = "\"$(missing_indices[1])\""
+            throw(ArgumentError(string(
+                "Index ", missing_string, " is occuring only on left side")))
+        else
+            missing_string = join(("\"$ix\"" for ix in missing_indices),
+                                  ", ", " and ")
+            throw(ArgumentError(string(
+                "Indices ", missing_string, " are occuring only on left side")))
+        end
+    end
+end
 
-function nest_loops(ex::Expr, idx::Vector{Symbol}, dim::Vector{Expr}, simd::Bool, threads::Bool)
-    isempty(idx) && return ex
+
+"""
+    nest_loops(expr, indices, axis_exprs, simd, threads) -> Expr
+
+Construct a nested loop around `expr`, using `indices` in ranges `axis_exprs`.
+
+# Example
+```julia-repl
+julia> nest_loops(:(A[i] = B[i]), [:i], [:(size(A, 1))], true, false)
+quote
+    local i
+    @simd for i = 1:size(A, 1)
+        A[i] = B[i]
+    end
+end
+```
+"""
+function nest_loops(expr::Expr,
+                    index_names::Vector{Symbol}, axis_expressions::Vector{Expr},
+                    simd::Bool, threads::Bool)
+    isempty(index_names) && return expr
 
     # Add @simd to the innermost loop, if required
     # and @threads to the outermost loop
-    ex = nest_loop(ex, idx[1], dim[1], simd, threads && 1==length(idx))
+    expr = nest_loop(expr, index_names[1], axis_expressions[1],
+                     simd, threads && length(index_names) == 1)
 
     # Add remaining for loops
-    for j = 2:length(idx)
-        ex = nest_loop(ex, idx[j], dim[j], false, threads && j==length(idx))
+    for j = 2:length(index_names)
+        expr = nest_loop(expr, index_names[j], axis_expressions[j],
+                         false, threads && length(index_names) == j)
     end
 
-    return ex
+    return expr
 end
 
-function nest_loop(ex::Expr, ix::Symbol, dim::Expr, simd::Bool, threads::Bool)
-    loop = :(for $ix = 1:$dim
-                 $ex
+function nest_loop(expr::Expr, index_name::Symbol, axis_expression::Expr,
+                   simd::Bool, threads::Bool)
+    loop = :(for $index_name = 1:$axis_expression
+                 $expr
              end)
 
     if threads
-      loop = :(Threads.@threads $loop)
+        loop = :(Threads.@threads $loop)
     elseif simd
-      loop = :(@simd $loop)
+        loop = :(@simd $loop)
     end
 
     return quote
-        local $ix
+        local $index_name
         $loop
     end
```

Comment:
I'm pretty sure we can just `return loop` here -- the for loop automatically makes its index local.